exercises/react/example.cr

```diff
@@ -0,0 +1,72 @@
+class Cell
+  @dependents = [] of ComputeCell
+
+  def initialize(initial : Int32)
+    @value = initial
+  end
+
+  getter :value
+
+  protected getter :dependents
+end
+
+class InputCell < Cell
+  def value=(new_value : Int32)
+    @value = new_value
+    @dependents.each(&.propagate)
+    @dependents.each(&.fire_callbacks)
+  end
+end
+
+class ComputeCell < Cell
+  alias CallbackID = Int32
+
+  @callbacks_given = 0
+  @callbacks = {} of CallbackID => Int32 ->
+  @last_callback_value : Int32
+
+  def initialize(new_value : -> Int32, *inputs : Cell)
+    super(new_value.call)
+    @new_value = new_value
+    @last_callback_value = @value
+    inputs.each { |i| i.dependents << self }
+  end
+
+  def add_callback(&block : Int32 ->) : CallbackID
+    @callbacks_given += 1
+    @callbacks[@callbacks_given] = block
+    @callbacks_given
+  end
+
+  def remove_callback(id : CallbackID)
+    @callbacks.delete(id)
+  end
+
+  def propagate
+    v = @new_value.call
+    return if v == @value
+    @value = v
+    @dependents.each(&.propagate)
+  end
+
+  def fire_callbacks
+    return if @value == @last_callback_value
+    @last_callback_value = @value
+    @callbacks.each_value { |c| c.call(@value) }
+    @dependents.each(&.fire_callbacks)
+  end
+end
+
+class Reactor
```

Comment:
Okay, I changed my mind about it because it's always possible that someone can do `typeof(existing_cell).new(some_args)`, and if they can do that then I want to make sure `initialize` has to be called with sensible args, and I prefer to do that in one place rather than two.